simulator/func_sim/func_sim.cpp

```diff
@@ -21,9 +21,20 @@ std::string MIPS::step()
     // fetch
     uint32 instr_bytes = mem->fetch( PC);
 
-    // decode
-    FuncInstr instr( instr_bytes, PC);
 
+    // check whether an instruction has been already decoded
+    // and decode it if it is not in the cache
+    auto it = instr_cache.find( instr_bytes);
```

Comment:
No, no, you did not get the idea.

We are finding an instruction by its PC, not by bytes.

Reply:
Is it the way all implementations of a cache for decoded instructions should work? I mean, it's only useful in loops when PC is used as a key. I don't have a problem implementing it, I just want to know why this approach is better.

Reply:
Fact 1: average program spends 80% of its time while executing loops.
Fact 2: by using PC as a key, we bypass a memory read, which should be as costly as decoding (decoding is just an arithmetic and looking up a small cached table)

Reply:
OK, we do not bypass memory read, as we add another one, but it still makes sense I believe

Reply:
Should I check the cache before fetching from the memory? There will be a problem with self-modifying code if I do so.

Reply:
I think the best move would be inclusion of the cache into MIPSMemory, so two methods will be updated:
* `FuncInstr MIPSMemory::fetch_instr( Addr PC)` will return instruction from cache or memory and update cache
* `void store( const FuncInstr& instr)` will _snoop_ the cache (ie invalidate the entry which is no longer valid)

Please put your files to `infra/instrcache`.